src/typechecker.erl

```diff
@@ -762,6 +762,13 @@ normalize_rec({type, _, union, Tys} = Type, Env, Unfolded) ->
                 Ts  -> type(union, Ts)
             end
     end;
+normalize_rec({type, _, list, [ElemTy]} = Type, Env, Unfolded) ->
+    case maps:get(mta(Type, Env), Unfolded, no_type) of
+        {type, NormType} -> NormType;
+        no_type ->
+            UnfoldedNew = maps:put(mta(Type, Env), {type, Type}, Unfolded),
+            type(list, [normalize_rec(typelib:remove_pos(ElemTy), Env, UnfoldedNew)])
+    end;
```

Comment:
Please don't recurse into the element type. (I don't think we should recurse into tuple, map and record elements either. If we do, I think it's the wrong place to do it.)

Also don't add `remove_pos`. It is recursive on the syntactic level and should only be done when a type comes into the system (from the parser or from gradualizer_db). I'm working on a small fix to remove all redundant remove_pos calls.

Doesn't the refinement work without this?

Reply:
> Doesn't the refinement work without this?

It doesn't in the current shape of the PR :( I was specifically trying that out after your statement about `normalize` (https://github.com/josefs/Gradualizer/pull/359#issuecomment-1025886673):

> Yes, I believe making normalize recursive was a mistake. It was made recursive in 766ba27 [...]

I was thinking about it a bit. This PR fixes regressions introduced when extending the exhaustiveness checking mechanism to non-trivial types. `normalize` was already recursive when I started that work. This PR makes regressions which were evident when running `make gradualize` finally go away. Therefore, I propose merging it without first reverting to non-recursive `normalize`. `normalize` is already recursive on `master`, this PR fixes known issues and doesn’t break anything in return, so I think it’s a reasonable step.

Then, I’d like to focus on getting back to recursive type problems. At this point it might turn out it’s actually best to return `normalize` to being non-recursive (or non-recursive where possible) and push subsequent `normalize` calls down where they’re actually needed. However, I think it might be quite a significant effort, so I think it makes sense to make it a next step, given the work in this PR already brings visible improvement. In other words: let's chunk the work.